File: 2529-range-product-queries-of-powers/2529-range-product-queries-of-powers.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
#include <stdio.h>
#include <stdlib.h>
#define MOD 1000000007

int* productQueries(int n, int** queries, int queriesSize, int* queriesColSize, int* returnSize) {
    int powers[32];
    
    int index = 0;
    long long temp = 1;
    int temp_n = n; 
    
    while (temp_n) {
        if (temp_n == 1){
            powers[index++] = 1;
            break; 
        }
        if (temp > temp_n){
            temp /= 2;
            temp_n -= (int)temp;
            powers[index++] = (int)temp;
            temp = 1;
        } else {
            temp *= 2;
        }
    }

    for (int i = 0; i < index / 2; i++){
        int tmp = powers[i];
        powers[i] = powers[index - 1 - i];
        powers[index - 1 - i] = tmp;
    }

    int* answer = (int*)malloc(sizeof(int) * queriesSize);
    *returnSize = queriesSize;
    
    for (int i = 0; i < queriesSize; i++){
        int x = queries[i][0];
        int y = queries[i][1];

        long long prod = 1;
        for (int j = x; j <= y; j++) {
            prod = (prod * powers[j]) % MOD;
        }
        answer[i] = (int)prod;
    }

    return answer;
}

File: 2529-range-product-queries-of-powers/2529-range-product-queries-of-powers.py
import math
class Solution:
    def productQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        powers = []
        temp = 1
        answer = []
        while n:
            if n == 1:
                powers.append(1)
                break 
            if temp > n:
                temp /= 2
                n -= temp
                powers.append(int(temp))
                temp = 1
            else:
                temp *= 2
        powers.reverse()
        
        for x, y in queries:
            if x == y:
                answer.append(powers[x] % (10**9 + 7))
            else:
                answer.append(math.prod(powers[x:y+1]) % (10**9 + 7))
        return answer
            